src/core/ext/filters/client_channel/xds/xds_api.cc

```diff
@@ -279,183 +252,106 @@ grpc_slice XdsLdsRequestCreateAndEncode(const std::string& server_name,
-                                        grpc_error* error) {
+}  // namespace
+
+grpc_slice XdsApi::CreateUnsupportedTypeNackRequest(const std::string& type_url,
+                                                    const std::string& nonce,
+                                                    grpc_error* error) {
+  upb::Arena arena;
+  envoy_api_v2_DiscoveryRequest* request = CreateDiscoveryRequest(
+      arena.ptr(), type_url.c_str(), /*version=*/"", nonce, error,
+      /*node=*/nullptr, /*build_version=*/nullptr);
+  return SerializeDiscoveryRequest(arena.ptr(), request);
+}
+
+grpc_slice XdsApi::CreateLdsRequest(const std::string& server_name,
+                                    const std::string& version,
+                                    const std::string& nonce, grpc_error* error,
+                                    bool populate_node) {
   upb::Arena arena;
-  // Create a request.
   envoy_api_v2_DiscoveryRequest* request =
-      envoy_api_v2_DiscoveryRequest_new(arena.ptr());
-  // Set version_info.
-  if (!version.empty()) {
-    envoy_api_v2_DiscoveryRequest_set_version_info(
-        request, upb_strview_makez(version.c_str()));
-  }
-  // Populate node.
-  if (build_version != nullptr) {
-    envoy_api_v2_core_Node* node_msg =
-        envoy_api_v2_DiscoveryRequest_mutable_node(request, arena.ptr());
-    PopulateNode(arena.ptr(), node, build_version, node_msg);
-  }
+      CreateDiscoveryRequest(arena.ptr(), kLdsTypeUrl, version, nonce, error,
+                             populate_node ? node_ : nullptr,
+                             populate_node ? build_version_ : nullptr);
+  // Add resource_name.
+  envoy_api_v2_DiscoveryRequest_add_resource_names(
+      request, upb_strview_make(server_name.data(), server_name.size()),
+      arena.ptr());
+  return SerializeDiscoveryRequest(arena.ptr(), request);
+}
+
+grpc_slice XdsApi::CreateRdsRequest(const std::string& route_config_name,
+                                    const std::string& version,
+                                    const std::string& nonce, grpc_error* error,
+                                    bool populate_node) {
+  upb::Arena arena;
+  envoy_api_v2_DiscoveryRequest* request =
+      CreateDiscoveryRequest(arena.ptr(), kRdsTypeUrl, version, nonce, error,
+                             populate_node ? node_ : nullptr,
+                             populate_node ? build_version_ : nullptr);
   // Add resource_name.
   envoy_api_v2_DiscoveryRequest_add_resource_names(
       request,
       upb_strview_make(route_config_name.data(), route_config_name.size()),
       arena.ptr());
-  // Set type_url.
-  envoy_api_v2_DiscoveryRequest_set_type_url(request,
-                                             upb_strview_makez(kRdsTypeUrl));
-  // Set nonce.
-  if (!nonce.empty()) {
-    envoy_api_v2_DiscoveryRequest_set_response_nonce(
-        request, upb_strview_makez(nonce.c_str()));
-  }
-  // Set error_detail if it's a NACK.
-  if (error != GRPC_ERROR_NONE) {
```

Comment:
Maybe i am missing something, but i feel like from 329-341 this error handling is not included in the refactored code?

Reply:
should it be added to CreateDiscoveryRequest ?

Reply:
As per our discussion, it is actually present in `CreateDiscoveryRequest()`.  It's just that the way github displays diffs makes it a little non-obvious.